app/src/main/java/com/arctouch/codechallenge/data/MoviesRepository.kt

```diff
@@ -0,0 +1,59 @@
+package com.arctouch.codechallenge.data
+
+import com.arctouch.codechallenge.model.Movie
+
+class MoviesRepository private constructor(
+    private val moviesRemoteDataSource: MoviesDataSource
+): MoviesDataSource {
+
+    companion object {
+        private var INSTANCE: MoviesRepository? = null
+
+        @JvmStatic
+        fun getInstance(remoteDataSource: MoviesDataSource): MoviesRepository {
+            return INSTANCE
+                ?: MoviesRepository(remoteDataSource).apply { INSTANCE = this }
+        }
+    }
+
+    override fun getMovies(callback: MoviesDataSource.LoadMoviesCallback, page: Long) {
+        if (Cache.cacheIsDirty) {
+            getMoviesFromRemoteSource(callback, page)
+        } else {
+            callback.onMoviesLoaded(Cache.movies)
+        }
+    }
+
+    private fun getMoviesFromRemoteSource(callback: MoviesDataSource.LoadMoviesCallback, page: Long) {
+        moviesRemoteDataSource.getMovies(object : MoviesDataSource.LoadMoviesCallback {
+            override fun onMoviesLoaded(movies: List<Movie>) {
+                Cache.cacheMovies(movies)
+                callback.onMoviesLoaded(Cache.movies)
+            }
+
+            override fun onDataNotAvailable() {
+                callback.onDataNotAvailable()
+            }
+        }, page)
+    }
+
+    override fun getMovie(movieId: Long, callback: MoviesDataSource.GetMovieCallback) {
+        if (Cache.movie?.id == movieId) {
+            callback.onMovieLoaded(Cache.movie!!)
```

Comment:
Please, avoid the usage of ```!!```. You can always use ```let```

An exemple:

```
override fun getMovie(movieId: Long, callback: MoviesDataSource.GetMovieCallback) {
        Cache.movie?.let {
            if (it.id == movieId) callback.onMovieLoaded(it)
        } ?: moviesRemoteDataSource.getMovie(movieId, callback)
    }
```

Reply:
You're right, I think I used `let` in other places but forgot it here.
Using "elvis operator" also makes it more elegant.

Reply:
Actually, I've just noticed that your example does not work in this case, because it doesn't call `moviesRemoteDataSource` when `Cache.movie` is not null but is different from requested `movieId`.

To fix it we could use something like that:
```
Cache.movie?.let {
    return@let if (it.id == movieId) { callback.onMovieLoaded(it) } else null
} ?: moviesRemoteDataSource.getMovie(movieId, callback)
```
But I think its readability is really poor, so I tried some other solutions and updated it with the one I thought was more readable. You can check [commit #6e8c88](https://github.com/diogojs/tmdb_app/pull/1/commits/6e8c8884763906025350f9aa9c64b5617d2758cf).